src/gc.c

```diff
@@ -1,12 +1,114 @@
 // This file is a part of Julia. License is MIT: https://julialang.org/license
 
 #include "gc.h"
+#include "julia_gcext.h"
 #include "julia_assert.h"
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
+// Linked list of callback functions
+
+typedef void (*jl_gc_cb_func_t)(void);
+
+typedef struct jl_gc_callback_list_t {
+    struct jl_gc_callback_list_t *next;
+    jl_gc_cb_func_t func;
+} jl_gc_callback_list_t;
+
+static jl_gc_callback_list_t *gc_cblist_root_scanner;
+static jl_gc_callback_list_t *gc_cblist_task_scanner;
+static jl_gc_callback_list_t *gc_cblist_pre_gc;
+static jl_gc_callback_list_t *gc_cblist_post_gc;
+static jl_gc_callback_list_t *gc_cblist_notify_external_alloc;
+static jl_gc_callback_list_t *gc_cblist_notify_external_free;
+
+#define gc_invoke_callbacks(kind, args) \
```

Comment:
So, just to clarify: you prefer that we replicate this code 6 times? The justification being that this way one can grep for e.g. `gc_cblist_root_scanner` to find where it is used?

A simple alternative to achieve that would be to add a comment to each of the 6 invocations of `gc_invoke_callbacks`, e.g.
```C
    // iterate over gc_cblist_root_scanner
    gc_invoke_callbacks(notify_external_alloc, (v, allocsz));
```

Reply:
> So, just to clarify: you prefer that we replicate this code 6 times?

That'll be better, yes. But you don't have to, The only bad thing about this is the symbol concatenation for the variable and type names, just spell them out explicitly will be fine.

Reply:
Got it. I had seen your earlier comment, but thought that was about the public API only, which I changed accordingly (whereas this use is internal). This is easy enough to change, too, but to explain why I did it: I wanted to keep the code typesafe so that list name and type always match. This is not very critical, though, as this is only used in a handful of places and only in one file, so if avoiding symbol concatenation is the bigger priority that's also not a problem.

Reply:
Yes, I know type safety is the issue. There are enough of these in the GC code that I hope we could use c++ here but fortunately most of these could be solved without too much more typing and without relying on a lot of macro tricks so far.............

Reply:
Congratulations. That exact macro is the very reason I hate this kind of macro so much. Unless I actually want to change something about these macros I'm not a fan of just changing those but I also don't want to introduce new ones.

Reply:
OK, I understand now: you really want all gone, and the existing ones are "legacy"